### ISSUE TITLE
DNSPlugin for INWX.de

### ISSUE THREAD

**Issue by andreashaerter** [rmbolger/Posh-ACME#573]
It would be nice to add support for INWX DNS-API.

**Comparable plugins**

* [acme.sh](https://github.com/acmesh-official/acme.sh/blob/master/dnsapi/dns_inwx.sh)
* [certbot](https://github.com/oGGy990/certbot-dns-inwx)

**Documentation**

* https://www.inwx.de/en/offer/api
* https://www.inwx.de/en/help/apidoc, especially
  * [2.15. Nameserver](https://www.inwx.de/en/help/apidoc/f/ch02s15.html)
  * [2.15.3. nameserver.createRecord](https://www.inwx.de/en/help/apidoc/f/ch02s15.html#nameserver.createRecord)
  * [2.15.5. nameserver.deleteRecord](https://www.inwx.de/en/help/apidoc/f/ch02s15.html#nameserver.deleteRecord)
  * [2.15.8. nameserver.exportrecords](https://www.inwx.de/en/help/apidoc/f/ch02s15.html#nameserver.exportrecords)
  * [2.15.12. nameserver.updateRecord](https://www.inwx.de/en/help/apidoc/f/ch02s15.html#nameserver.updateRecord)

There is also an OT&E test system, which one can access via [ote.inwx.com](https://ote.inwx.com/en). It provides the usual WebUI and API using a test database. On the OTE system no actions will be charged. So one can test how to register domains etc., a OT&E account can be created at https://www.ote.inwx.de/en/customer/signup

Please let me know if I can assist by providing a real domain and API user.

**Comment by rmbolger** [comment]
Hey @andreashaerter, thanks for the super helpful info. It's so nice when providers have a test environment you can use to build against. Theoretically, that should work just fine.

I'll put this in my queue to work on when I get a chance.

**Comment by andreashaerter** [comment]
@rmbolger please let me know if / when you started developing (just to prevent double work) as I might find some time to read your documentation at https://poshac.me/docs/v4/Plugins/Plugin-Development-Guide/ and create a PR within the next weeks. I will leave a comment here if so.

BTW: The whole project's quality and documentation is impressive.

**Comment by rmbolger** [comment]
Will do. I haven't started anything yet though.

> BTW: The whole project's quality and documentation is impressive.

Thanks! It's nice for the effort to be recognized.

**Comment by andreashaerter** [comment]
Hi @rmbolger, I just wanted to say I will start developing now.

**Comment by rmbolger** [comment]
Sounds good. Looking forward to the result.

**Comment by andreashaerter** [comment]
I finished a working Plugin using the JSON RPC-API, will get some sleep and test it in some real-world deployments and open a PR then.

Small question: INWX supports a shared secret to generate 2FA TOTP codes. I did not implement this yet (but prepared the parameter + an exception like 

```ps
[...]
    .PARAMETER INWXSharedSecret
        FIXME to be implemented, patches welcome.

        If your account is secured by mobile TAN ("2FA", "2 factor authentication"),
        you also have to define the shared secret (usually presented below the QR
        code during mobile TAN setup) to enable this function to generate OTP codes.
        The shared secret is NOT not the 6 digit code you need to enter when logging
        in. If you are not using 2FA, do not define this parameter.
[....]

    # generate needed OTP
    if ($INWXSharedSecret) {
        # get shared secret as plaintext
        $INWXSharedSecretInsecure = [pscredential]::new('a',$INWXSharedSecret).GetNetworkCredential().Password

        # FIXME to be implemented
        # Propably useful: https://github.com/acmesh-official/acme.sh/blob/master/dnsapi/dns_inwx.sh#L222C4-L263
        throw "Sorry, the INWX Shared Secret / 2FA functionality is not supported yet. Patches are welcome."
    }
```

is this OK for you for a working 1.0? And: Is there some already existing OTP infra existing (to be used by plugins) or would an inline helper be the best way to create the OTP secrets?

**Comment by rmbolger** [comment]
No existing plugins implement any sort of TOTP functionality that I recall.

Using MFA for automated processes seems like it would defeat the purpose of either MFA or being automated. If the process can provide the MFA code by having the seed stored with the credentials, it's not really MFA anymore. It's just an extra first factor password. So why bother? And if it requires human interaction to provide the code, it's no longer fully automated.

Bottom line, it would be totally acceptable to not provide any sort of MFA compatibility and add a warning in the guide that the plugin won't work with MFA-enabled accounts.

**Comment by andreashaerter** [comment]
Fine. I will finish testing the first version then.

However, implementing OTP makes sense to a certain extent: INWX provides the ability to create multiple sub-accounts and has a fixed role model, including a 'DNS manager' role, which only allows altering DNS records—ideal for tasks like ACME-related automation. The flaw is that 2FA is a global setting, influencing not only the main account but all accounts, including the unprivileged ones, as well as the API access, which would also require sending an OTP.

**Comment by rmbolger** [comment]
> However, implementing OTP makes sense to a certain extent

My apologies. I didn't mean to imply you *shouldn't* implement the TOTP functionality. I just didn't want you to think you *had* to for it to be considered done.

**Comment by andreashaerter** [comment]
> > However, implementing OTP makes sense to a certain extent
> 
> My apologies. I didn't mean to imply you _shouldn't_ implement the TOTP functionality. I just didn't want you to think you _had_ to for it to be considered done.

:-D No worries! I didn't think of it that way, and it wasn’t my intention to sound harsh. Maybe it's a nuance that a non-native speaker might not catch.

Let me know if you'd like any further adjustments!

**Comment by andreashaerter** [comment]
Solved by https://github.com/rmbolger/Posh-ACME/pull/585
